src/box/errcode.h

```diff
@@ -234,7 +234,7 @@ struct errcode_record {
 	_(ER_PRIV_NOT_GRANTED, 91,		"User '%s' does not have %s access on %s '%s'", "user", STRING, "privilege", STRING, "object_type", STRING, "object", STRING) \
 	_(ER_ROLE_NOT_GRANTED, 92,		"User '%s' does not have role '%s'", "user", STRING, "role", STRING) \
 	_(ER_MISSING_SNAPSHOT, 93,		"Can't find snapshot") \
-	_(ER_CANT_UPDATE_PRIMARY_KEY, 94,	"Attempt to modify a tuple field which is part of primary index in space '%s'", "space", STRING) \
+	_(ER_CANT_UPDATE_PRIMARY_KEY, 94,	"Attempt to modify a tuple field which is part of primary index in space '%s'", "space", STRING, "index", STRING, "old_tuple", TUPLE, "new_tuple", TUPLE) \
```

Comment:
It'd be nice to store space and index ids in this error as well - for consistency with other errors.

Reply:
Done.

<details>
<summary>diff</summary>

```diff
diff --git a/src/box/errcode.h b/src/box/errcode.h
index 70e9f282d..aec1bc82c 100644
--- a/src/box/errcode.h
+++ b/src/box/errcode.h
@@ -234,7 +234,7 @@ struct errcode_record {
 	_(ER_PRIV_NOT_GRANTED, 91,		"User '%s' does not have %s access on %s '%s'", "user", STRING, "privilege", STRING, "object_type", STRING, "object", STRING) \
 	_(ER_ROLE_NOT_GRANTED, 92,		"User '%s' does not have role '%s'", "user", STRING, "role", STRING) \
 	_(ER_MISSING_SNAPSHOT, 93,		"Can't find snapshot") \
-	_(ER_CANT_UPDATE_PRIMARY_KEY, 94,	"Attempt to modify a tuple field which is part of primary index in space '%s'", "space", STRING, "index", STRING, "old_tuple", TUPLE, "new_tuple", TUPLE) \
+	_(ER_CANT_UPDATE_PRIMARY_KEY, 94,	"Attempt to modify a tuple field which is part of primary index in space '%s'", "space", STRING, "space_id", UINT, "index", STRING, "index_id", UINT, "old_tuple", TUPLE, "new_tuple", TUPLE) \
 	_(ER_UPDATE_INTEGER_OVERFLOW, 95,	"Integer overflow when performing '%c' operation on field %s", "operation", CHAR, "field", STRING) \
 	_(ER_GUEST_USER_PASSWORD, 96,		"Setting password for guest user has no effect") \
 	_(ER_TRANSACTION_CONFLICT, 97,		"Transaction has been aborted by conflict") \
diff --git a/src/box/index.cc b/src/box/index.cc
index 96a8e0d8b..438f5f5ef 100644
--- a/src/box/index.cc
+++ b/src/box/index.cc
@@ -221,7 +221,8 @@ index_check_dup(struct index *index, struct tuple *old_tuple,
 			struct space *space = space_by_id(index->def->space_id);
 			assert(space != NULL);
 			diag_set(ClientError, ER_CANT_UPDATE_PRIMARY_KEY,
-				 space->def->name, index->def->name,
+				 space->def->name, space->def->id,
+				 index->def->name, index->def->iid,
 				 old_tuple, new_tuple);
 			return -1;
 		}
diff --git a/src/box/memtx_space.c b/src/box/memtx_space.c
index 47e973fdf..613b8f27e 100644
--- a/src/box/memtx_space.c
+++ b/src/box/memtx_space.c
@@ -656,7 +656,8 @@ memtx_space_execute_upsert(struct space *space, struct txn *txn,
 				  HINT_NONE, pk->def->key_def) != 0) {
 			/* Primary key is changed: log error and do nothing. */
 			diag_set(ClientError, ER_CANT_UPDATE_PRIMARY_KEY,
-				 space_name(space), pk->def->name,
+				 space_name(space), space_id(space),
+				 pk->def->name, pk->def->iid,
 				 old_tuple, new_tuple);
 			diag_log();
 			tuple_unref(new_tuple);
diff --git a/src/box/session_settings.c b/src/box/session_settings.c
index d94a9470c..420145ff7 100644
--- a/src/box/session_settings.c
+++ b/src/box/session_settings.c
@@ -373,8 +373,8 @@ session_settings_space_execute_update(struct space *space, struct txn *txn,
 			struct tuple *old_tuple =
 				tuple_new(format, old_data, old_data_end);
 			diag_set(ClientError, ER_CANT_UPDATE_PRIMARY_KEY,
-				 space_name(space), pk_def->name,
-				 old_tuple, *result);
+				 space_name(space), space_id(space),
+				 pk_def->name, pk_def->iid, old_tuple, *result);
 			tuple_delete(old_tuple);
 			goto finish;
 		}
diff --git a/src/box/space.c b/src/box/space.c
index c6131643b..a1c793565 100644
--- a/src/box/space.c
+++ b/src/box/space.c
@@ -1284,8 +1284,8 @@ after_old_tuple_lookup:;
 	    tuple_compare(old_tuple, HINT_NONE, new_tuple, HINT_NONE,
 			  pk->def->key_def) != 0) {
 		diag_set(ClientError, ER_CANT_UPDATE_PRIMARY_KEY,
-			 space->def->name, index->def->name,
-			 old_tuple, new_tuple);
+			 space->def->name, space->def->id, index->def->name,
+			 index->def->iid, old_tuple, new_tuple);
 		rc = -1;
 		goto out;
 	}
diff --git a/src/box/vinyl.c b/src/box/vinyl.c
index 6cc09d217..c875a2773 100644
--- a/src/box/vinyl.c
+++ b/src/box/vinyl.c
@@ -1798,8 +1798,8 @@ vy_check_update(struct space *space, const struct vy_lsm *pk,
 	    vy_stmt_compare(old_tuple, HINT_NONE, new_tuple,
 			    HINT_NONE, pk->key_def) != 0) {
 		diag_set(ClientError, ER_CANT_UPDATE_PRIMARY_KEY,
-			 space_name(space), pk->base.def->name,
-			 old_tuple, new_tuple);
+			 space_name(space), space_id(space), pk->base.def->name,
+			 pk->base.def->iid, old_tuple, new_tuple);
 		return -1;
 	}
 	return 0;
diff --git a/test/box-luatest/error_subsystem_improvements_test.lua b/test/box-luatest/error_subsystem_improvements_test.lua
index 823a6943d..bbb559fd5 100644
--- a/test/box-luatest/error_subsystem_improvements_test.lua
+++ b/test/box-luatest/error_subsystem_improvements_test.lua
@@ -817,7 +817,9 @@ g.test_dml_update_primary_key = function(cg)
         test:replace({1, 2})
         t.assert_error_covers({
             space = 'test',
+            space_id = test.id,
             index = 'primary',
+            index_id = 0,
             old_tuple = {1, 2},
             new_tuple = {3, 2},
             message = "Attempt to modify a tuple field which is part of " ..
@@ -830,7 +832,9 @@ g.test_dml_update_primary_key = function(cg)
         test:before_replace(update_pk)
         t.assert_error_covers({
             space = 'test',
+            space_id = test.id,
             index = 'primary',
+            index_id = 0,
             old_tuple = {1, 2},
             new_tuple = {2, 3},
             message = "Attempt to modify a tuple field which is part of " ..
@@ -843,7 +847,9 @@ g.test_dml_update_primary_key = function(cg)
         test:replace({1, 2})
         t.assert_error_covers({
             space = 'test',
+            space_id = test.id,
             index = 'primary',
+            index_id = 0,
             old_tuple = {1, 2},
             new_tuple = {3, 2},
             message = "Attempt to modify a tuple field which is part of " ..
@@ -855,7 +861,9 @@ g.test_dml_update_primary_key = function(cg)
         space:update('sql_full_metadata', {{'=', 2, false}})
         t.assert_error_covers({
             space = '_session_settings',
+            space_id = space.id,
             index = 'primary',
+            index_id = 0,
             old_tuple = {'sql_full_metadata', false},
             new_tuple = {'foo', false},
             message = "Attempt to modify a tuple field which is part of " ..
@@ -867,7 +875,9 @@ g.test_dml_update_primary_key = function(cg)
         test:replace({1, 2})
         t.assert_error_covers({
             space = 'test',
+            space_id = test.id,
             index = 'primary',
+            index_id = 0,
             old_tuple = {1, 2},
             new_tuple = {3, 2},
             message = "Attempt to modify a tuple field which is part of " ..

```

</details>